mediators/datamapper/components/org.wso2.carbon.gateway.mediators.datamapper.engine/src/main/java/org/wso2/carbon/gateway/mediators/datamapper/engine/core/schemas/JacksonJSONSchema.java

```diff
@@ -0,0 +1,375 @@
+                String schemaType = getSchemaType(subSchema);
+                if (ARRAY_ELEMENT_TYPE.equals(schemaType)) {
+                    elementType = getElementTypeByName(elementName, getSchemaItems(subSchema));
+                } else {
+                    elementType = getElementTypeByName(elementName, subSchema);
+                }
+                if (elementType != null) {
+                    return elementType;
+                }
+            }
+        }
+        return null;
+    }
+
+    private Map<String, Object> getSchemaProperties(Map<String, Object> schema) {
+        Map<String, Object> nextSchema = new HashMap<>();
+        if (schema.containsKey(PROPERTIES_KEY)) {
+            nextSchema.putAll((Map<? extends String, Object>) schema.get(PROPERTIES_KEY));
+        }
+        if (schema.containsKey(ATTRIBUTES_KEY)) {
+            nextSchema.putAll((Map<? extends String, Object>) schema.get(ATTRIBUTES_KEY));
+        }
+        return nextSchema;
+    }
+
+    /**
+     * Get all the items in the schema as a map.
+     * 
+     * @param schema    Schema
+     * @return          Schema as a map
+     */
+    public Map<String, Object> getSchemaItems(Map<String, Object> schema) {
+        Map<String, Object> nextSchema = new HashMap<>();
+        if (schema.containsKey(ITEMS_KEY)) {
+            Object propertyList = schema.get(ITEMS_KEY);
+            if (propertyList instanceof Map) {
+                nextSchema.putAll((Map<? extends String, Object>) propertyList);
+                if (nextSchema.containsKey(ATTRIBUTES_KEY)) {
+                    nextSchema.putAll((Map<? extends String, Object>) nextSchema.get(ATTRIBUTES_KEY));
+                }
+            } else {
+                nextSchema.putAll((Map<? extends String, Object>) ((ArrayList) propertyList).get(0));
+            }
+        } else {
+            throw new IllegalArgumentException("Given schema does not contain value under key : " + ITEMS_KEY);
+        }
+        if (schema.containsKey(ATTRIBUTES_KEY)) {
+            nextSchema.putAll((Map<? extends String, Object>) schema.get(ATTRIBUTES_KEY));
+        }
+        return nextSchema;
+    }
+
+    private String getSchemaType(Map<String, Object> schema) {
+        if (schema.containsKey(TYPE_KEY)) {
+            Object type = schema.get(TYPE_KEY);
+            if (type instanceof String) {
+                return (String) type;
+            } else {
+                throw new IllegalArgumentException("Illegal format " + type.getClass() + " value found under key : " +
+                        TYPE_KEY);
+            }
+        } else {
+            throw new IllegalArgumentException("Given schema does not contain value under key : " + TYPE_KEY);
+        }
+    }
+
+    @Override 
+    public String getPrefixForNamespace(String url) {
+        if (EMPTY_STRING.equals(url)) {
+            return EMPTY_STRING;
+        } else if (namespaceMap.containsKey(url)) {
```

Comment:
any reason as to why for one case we are returning EMPTY_STRING while in another case null? 
maybe its what we want to do (if that's the case it woud be nice to have it documented, otherwise 2 years down the line someone will be scratching our heads why? ) :) :) :)